### PR TITLE
Eliminate redundant distinct inside IN/NOT IN/EXISTS/NOT EXISTS subquery

### PR DESCRIPTION
**Pull request by hsyuan**
In an IN, EXISTS, NOT IN or NOT EXISTS subquery, any duplicates in the
subselect will not affect the overall result, so we can throw away any
DISTINCT clause. Unless there's a LIMIT.

### CODE REVIEW

File: libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 1]
```diff
@@ -2108,6 +2108,71 @@ CExpressionPreprocessor::PexprExistWithPredFromINSubq
...
+	GPOS_ASSERT(NULL != pmp);
+	GPOS_ASSERT(NULL != pexpr);
+
+	COperator *pop = pexpr->Pop();
+
+	bool fAnyAllSubq = COperator::EopScalarSubqueryAny == pop->Eopid() ||
+			COperator::EopScalarSubqueryAll == pop->Eopid();
+	bool fExistsSubq = COperator::EopScalarSubqueryExists == pop->Eopid() ||
+			 COperator::EopScalarSubqueryNotExists == pop->Eopid();
+
```
Comment (dhanashreek89) [inline libgpopt/src/operators/CExpressionPreprocessor.cpp]: Can be simplified if we use `CUtils::FQuantifiedSubquery(pop)` and `CUtils::FExistentialSubquery(pop)` for Any/All and Exists/NotExists respectively.

Comment (hsyuan) [reply libgpopt/src/operators/CExpressionPreprocessor.cpp]: Ah, yeah, thanks. Will do.

**Comment by oarap** [comment]
You need to process children first. Otherwise, you will only process top level distinct.

Look at the example below.

`explain select * from foo where exists (select distinct a from foo where exists (select distinct a from foo) );`
```
Algebrized query:
+--CLogicalSelect
   |--CLogicalGet "foo" ("foo"), Columns: ["a" (0), "b" (1), "c" (2), "ctid" (3), "xmin" (4), "cmin" (5), "xmax" (6), "cmax" (7), "tableoid" (8), "gp_segment_id" (9)] Key sets: {[3,9]}
   +--CScalarSubqueryExists
      +--CLogicalGbAgg( Global ) Grp Cols: ["a" (10)][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ]
         |--CLogicalSelect
         |  |--CLogicalGet "foo" ("foo"), Columns: ["a" (10), "b" (11), "c" (12), "ctid" (13), "xmin" (14), "cmin" (15), "xmax" (16), "cmax" (17), "tableoid" (18), "gp_segment_id" (19)] Key sets: {[3,9]}
         |  +--CScalarSubqueryExists
         |     +--CLogicalGbAgg( Global ) Grp Cols: ["a" (20)][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ]
         |        |--CLogicalGet "foo" ("foo"), Columns: ["a" (20), "b" (21), "c" (22), "ctid" (23), "xmin" (24), "cmin" (25), "xmax" (26), "cmax" (27), "tableoid" (28), "gp_segment_id" (29)] Key sets: {[3,9]}
         |        +--CScalarProjectList
         +--CScalarProjectList
Algebrized preprocessed query:
+--CLogicalSelect
   |--CLogicalGet "foo" ("foo"), Columns: ["a" (0), "b" (1), "c" (2), "ctid" (3), "xmin" (4), "cmin" (5), "xmax" (6), "cmax" (7), "tableoid" (8), "gp_segment_id" (9)] Key sets: {[3,9]}
   +--CScalarSubqueryExists
      +--CLogicalSelect
         |--CLogicalGet "foo" ("foo"), Columns: ["a" (10), "b" (11), "c" (12), "ctid" (13), "xmin" (14), "cmin" (15), "xmax" (16), "cmax" (17), "tableoid" (18), "gp_segment_id" (19)] Key sets: {[3,9]}
         +--CScalarSubqueryExists
            +--CLogicalGbAgg( Global ) Grp Cols: ["a" (20)][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ]
               |--CLogicalGet "foo" ("foo"), Columns: ["a" (20), "b" (21), "c" (22), "ctid" (23), "xmin" (24), "cmin" (25), "xmax" (26), "cmax" (27), "tableoid" (28), "gp_segment_id" (29)] Key sets: {[3,9]}
               +--CScalarProjectList
```

It should produce the tree below:
```
Algebrized preprocessed query:
+--CLogicalSelect
   |--CLogicalGet "foo" ("foo"), Columns: ["a" (0), "b" (1), "c" (2), "ctid" (3), "xmin" (4), "cmin" (5), "xmax" (6), "cmax" (7), "tableoid" (8), "gp_segment_id" (9)] Key sets: {[3,9]}
   +--CScalarSubqueryExists
      +--CLogicalSelect
         |--CLogicalGet "foo" ("foo"), Columns: ["a" (10), "b" (11), "c" (12), "ctid" (13), "xmin" (14), "cmin" (15), "xmax" (16), "cmax" (17), "tableoid" (18), "gp_segment_id" (19)] Key sets: {[3,9]}
         +--CScalarSubqueryExists
            +--CLogicalGet "foo" ("foo"), Columns: ["a" (20), "b" (21), "c" (22), "ctid" (23), "xmin" (24), "cmin" (25), "xmax" (26), "cmax" (27), "tableoid" (28), "gp_segment_id" (29)] Key sets: {[3,9]}
```

**Comment by hsyuan** [comment]
@oarap I addressed your comments, please take a look. Thanks!

**Pull request merged**